src/Stratis.Bitcoin/Utilities/AsyncExecutionEvent.cs

```diff
@@ -0,0 +1,202 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Stratis.Bitcoin.Utilities
+{
+    /// <summary>
+    /// Asynchronous event handler that can be registered with <see cref="AsyncExecutionEvent{TSender, TArg}/>.
+    /// </summary>
+    /// <typeparam name="TSender">Type of the sender object that is the source of the event.</typeparam>
+    /// <typeparam name="TArg">Type of the argument that is passed to the callback.</typeparam>
+    /// <param name="sender">Source of the event.</param>
+    /// <param name="arg">Callback argument.</param>
+    public delegate Task AsyncExecutionEventCallback<TSender, TArg>(TSender sender, TArg arg);
+
+    /// <summary>
+    /// Execution event is a specific moment in the execution flow of that a component 
+    /// that other components are allowed to be subscribed to and get notified about 
+    /// when it occurs.
+    /// <para>
+    /// This implementation allows components to register asynchronous event handlers.
+    /// </para>
+    /// </summary>
+    /// <typeparam name="TSender">Type of event source sender objects.</typeparam>
+    /// <typeparam name="TArg">Type of arguments that are passed to callbacks.</typeparam>
+    public class AsyncExecutionEvent<TSender, TArg>: IDisposable
+    {
+        /// <summary>
+        /// Protects access to <see cref="callbackList"/> and <see cref="callbackToListNodeMapping"/>,
+        /// and also provides gurantees of <see cref="Unregister(T)"/> method.
```

Comment:
typo. `guarantees`

also instead of `<see cref="Unregister(T)"/>` which T reference cannot be resolved use:
`<see cref="Unregister(AsyncExecutionEventCallback{TSender, TArg})"/>`

Reply:
excellent, tx